### PR TITLE
OCPBUGS-17249: UPSTREAM: <carry>: openshift-kube-apiserver: pod .spec.nodeName should not override project node selector in podNodeEnvironment admission plugin

### PR DESCRIPTION
**Pull request by atiratree**

### CODE REVIEW

File: openshift-kube-apiserver/admission/scheduler/nodeenv/admission.go [hunk 1]
```diff
@@ -62,6 +62,11 @@ func (p *podNodeEnvironment) admit(ctx context.Context, a admission.Attributes,
 		return nil
 	}
 
+	// NodeName has a precedence
+	if len(pod.Spec.NodeName) > 0 {
```
Comment (soltysh) [inline openshift-kube-apiserver/admission/scheduler/nodeenv/admission.go:66]: @atiratree are you sure we want to allow setting nodename explicitly on a pod? Isn't the goal of project node selectors to ensure we can restrict authors to running only on those explicitly pointed out nodes?

**Review by soltysh — CHANGES_REQUESTED** [review]

File: openshift-kube-apiserver/admission/scheduler/nodeenv/admission.go (hunk 1 above)
Comment (atiratree) [reply openshift-kube-apiserver/admission/scheduler/nodeenv/admission.go]: Yes you are right, scheduling should not take a precedence over project node selector restrictions.

I have updated the PR, to increase the validations.

**Comment by kasturinarra** [comment]
will add cherry-pick-approved label once other labels are present.

File: openshift-kube-apiserver/admission/scheduler/nodeenv/admission.go [hunk 2]
```diff
@@ -96,6 +99,16 @@ func (p *podNodeEnvironment) admit(ctx context.Context, a admission.Attributes,
 		return apierrors.NewForbidden(resource, name, fmt.Errorf("pod node label selector does not extend project node label selector"))
 	}
 
+	if len(pod.Spec.NodeName) > 0 && len(projectNodeSelector) > 0 {
+		node, err := p.nodeLister.Get(pod.Spec.NodeName)
+		if err != nil {
+			return apierrors.NewForbidden(resource, name, fmt.Errorf("cannot validate project node label selector: %v", err))
+		}
+		if !isSubset(projectNodeSelector, node.Labels) {
```
[the remark this answers is not in the document]
Comment (atiratree) [reply openshift-kube-apiserver/admission/scheduler/nodeenv/admission.go]: updated, and added a note about a possible future improvement

**Review by soltysh — APPROVED** [review]
/lgtm
/approve

**Pull request merged**